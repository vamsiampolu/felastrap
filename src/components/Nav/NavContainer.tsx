import type { CSSProperties } from "fela";
import type { ElementType } from "react";
import { useFela } from "react-fela";
import type { PolymorphicComponentPropsWithoutRef } from "../../polymorphicComponentProps";
import type { NavVariant, StyleProps } from "../common";
import { isTabs, isUnderline, isVertical } from "./utils";

export type CustomNavContainerProps = Partial<{
  card: boolean;
  center: boolean;
  fill: boolean;
  justify: boolean;
  variant: NavVariant;
}>;

export type NavContainerProps<C extends ElementType> =
  PolymorphicComponentPropsWithoutRef<C, CustomNavContainerProps>;

const rule = ({
  center = false,
  variant = "default",
}: StyleProps<CustomNavContainerProps>) => {
  const style: CSSProperties = {
    display: "flex",
    flexWrap: "wrap",
    paddingLeft: 0,
    marginBottom: 0,
    listStyle: "none",
  };

  if (center) {
    style.justifyContent = "center";
  } else if (isVertical({ variant })) {
    style.flexDirection = "column";
  }

  if (isUnderline({ variant })) {
    style.gap = "1rem";
  }

  if (isTabs({ variant })) {
    style.borderBottom = "1px solid #dee2e6";
  }

  return style;
};

export const NavContainer = <C extends ElementType = "ul">({
  className,
  center = false,
  variant = "default",
  as,
  children,
  ...other
}: NavContainerProps<C>) => {
  const { css } = useFela({ center, variant });
  const cx = `${css([rule])} nav-container ${className || ""}`;
  const Component = as || "ul";
  return (
    <Component data-testid="nav-container" className={cx} {...other}>
      {children}
    </Component>
  );
};
