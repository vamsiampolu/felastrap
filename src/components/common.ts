import type * as Css from "csstype";
import type * as Fela from "fela";
import type {
  ComponentPropsWithoutRef,
  ComponentType,
  ElementType,
} from "react";
import type * as ReactFela from "react-fela";

export type ExcludeFrom<T, U> = T extends U ? never : T;
export type Theme = object;

export type Rule = Fela.TRule;
export type CSSObject = Fela.CSSObject;
export type StyleProps<P = object> = ReactFela.StyleProps<Theme, P>;
export type StyleFunction<P = object> = ReactFela.StyleFunction<Theme, P>;

export type Padding = Required<
  Pick<
    Css.Properties,
    "paddingTop" | "paddingLeft" | "paddingRight" | "paddingBottom"
  >
>;
export type BorderRadius = Required<
  Pick<
    Css.Properties,
    | "borderTopLeftRadius"
    | "borderTopRightRadius"
    | "borderBottomRightRadius"
    | "borderBottomLeftRadius"
  >
>;

export type PaddingString = Required<Css.Properties["padding"]>;
export type Color = Css.Property.Color;

export type HeaderTypes = "h1" | "h2" | "h3" | "h4" | "h5" | "h6";
export type Size = "small" | "medium" | "large";
export type Variant =
  | "default"
  | "danger"
  | "info"
  | "primary"
  | "secondary"
  | "success"
  | "warning";

export type FontSize = Css.Property.FontSize;
export type ColorMapping = Record<Variant, Color>;
export type PaddingBySize = Record<Size, Padding>;

export type SpanProps = ComponentPropsWithoutRef<"span">;
export type DivProps = ComponentPropsWithoutRef<"div">;
export type ListItemProps = ComponentPropsWithoutRef<"li">;

export type DropdownDirection = "up" | "down" | "end" | "start" | "centered";
export type CaretDirection = "up" | "down" | "end" | "start";
export type NavVariant =
  | "default"
  | "tabs"
  | "pills"
  | "underline"
  | "vertical";

export type CaretProps = DivProps & {
  width: string | number | null;
};

export type DynamicElement<T extends ElementType = "span"> =
  | keyof JSX.IntrinsicElements
  | ComponentType<T>;
export type PargraphProps = ComponentPropsWithoutRef<"p">;
export type ImageProps = ComponentPropsWithoutRef<"img">;
export type TooltipPlacement = "top" | "left" | "bottom" | "right";

export type MediaQuerySize = "xs" | "sm" | "md" | "lg" | "xl" | "xxl";
export type Timer = ReturnType<typeof setTimeout>;
export type OnClick<T extends ElementType> =
  ComponentPropsWithoutRef<T>["onClick"];
export type StyleKey = Exclude<
  keyof HTMLElement["style"],
  "length" | "parentRule"
>;
