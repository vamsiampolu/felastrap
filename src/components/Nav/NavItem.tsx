import { useFela } from "react-fela";
import { ComponentPropsWithoutRef } from "react";
import { DynamicElement, StyleProps } from "../common";

export type CustomProps = Partial<{
  card: boolean;
  fill: boolean;
  justified: boolean;
  active: boolean;
  as: DynamicElement;
  className: string;
}>;
export type NavItemProps = ComponentPropsWithoutRef<"li"> & CustomProps;

const fillRule = ({ fill = false }: StyleProps<CustomProps>) =>
  fill
    ? {
        flex: "1 1 auto",
        textAlign: "center",
      }
    : {};

const justifiedRule = ({ justified = false }: StyleProps<CustomProps>) =>
  justified
    ? {
        flexBasis: 0,
        flexGrow: 1,
        textAlign: "center",
      }
    : {};
const cardRule = ({ card = false }: StyleProps<CustomProps>) => {
  if (!card) {
    return {};
  }

  return {
    marginLeft: "1rem",
  };
};

export const NavItem = ({
  children,
  card = false,
  className = "",
  fill = false,
  justified = false,
  active = false,
  ...other
}: NavItemProps) => {
  const { css } = useFela({ card, fill, justified, active });
  const cx = css([fillRule, justifiedRule, cardRule]);

  return (
    <li
      data-testid="nav-item"
      className={`nav-item ${cx} ${className || ""}`.trim()}
      {...other}
    >
      {children}
    </li>
  );
};
