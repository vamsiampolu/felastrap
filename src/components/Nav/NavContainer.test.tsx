import { IRenderer } from "fela";
import { NavContainer } from "./NavContainer";
import {
  Wrapper,
  createRenderer,
  createSnapshot,
  cssReboot,
} from "../../test-helpers";
import { NavItem } from "./NavItem";
import { render, screen } from "@testing-library/react";
import { NavVariant } from "../common";

describe("components/Nav/NavContainer", () => {
  let renderer: IRenderer;

  beforeEach(() => {
    renderer = createRenderer();
    cssReboot(renderer);
  });

  afterEach(() => {
    renderer.clear();
  });

  it("should render a NavContainer", async () => {
    const component = (
      <NavContainer>
        <NavItem>Hello</NavItem>
      </NavContainer>
    );

    const wrappedComponent = <Wrapper renderer={renderer}>{component}</Wrapper>;
    render(wrappedComponent);

    expect(screen.getByTestId("nav-container")).toBeInTheDocument();
    expect(screen.queryByTestId("nav-container")?.tagName).toBe("UL");

    const snapshot = await createSnapshot({ renderer, component });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render a centered NavContainer", async () => {
    const component = (
      <NavContainer center>
        <NavItem>Hello</NavItem>
      </NavContainer>
    );

    const wrappedComponent = <Wrapper renderer={renderer}>{component}</Wrapper>;
    render(wrappedComponent);

    expect(screen.getByTestId("nav-container")).toBeInTheDocument();
    expect(screen.queryByTestId("nav-container")?.tagName).toBe("UL");

    const snapshot = await createSnapshot({ renderer, component });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render a NavContainer as a nav", async () => {
    const component = (
      <NavContainer as="nav">
        <NavItem>Hello</NavItem>
      </NavContainer>
    );

    const wrappedComponent = <Wrapper renderer={renderer}>{component}</Wrapper>;
    render(wrappedComponent);

    expect(screen.getByTestId("nav-container")).toBeInTheDocument();
    expect(screen.queryByTestId("nav-container")?.tagName).toBe("NAV");

    const snapshot = await createSnapshot({ renderer, component });
    expect(snapshot).toMatchSnapshot();
  });

  it.each(["default", "tabs", "pills", "underline", "vertical"])(
    "should render a NavContainer of variant %s",
    async (variant) => {
      const component = (
        <NavContainer variant={variant as NavVariant}>
          <NavItem>Hello</NavItem>
        </NavContainer>
      );

      const wrappedComponent = (
        <Wrapper renderer={renderer}>{component}</Wrapper>
      );
      render(wrappedComponent);

      expect(screen.getByTestId("nav-container")).toBeInTheDocument();
      expect(screen.queryByTestId("nav-container")?.tagName).toBe("UL");

      const snapshot = await createSnapshot({ renderer, component });
      expect(snapshot).toMatchSnapshot();
    },
  );

  it.each(["default", "tabs", "pills", "underline"])(
    "should render a centered NavContainer of variant %s",
    async (variant) => {
      const component = (
        <NavContainer center variant={variant as NavVariant}>
          <NavItem>Hello</NavItem>
        </NavContainer>
      );

      const snapshot = await createSnapshot({ renderer, component });
      expect(snapshot).toMatchSnapshot();
    },
  );
});
