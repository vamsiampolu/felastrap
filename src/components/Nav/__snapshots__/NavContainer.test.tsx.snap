// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Nav/NavContainer should render a NavContainer 1`] = `
".a {
  display: flex;
}
.b {
  flex-wrap: wrap;
}
.c {
  padding-left: 0;
}
.d {
  margin-bottom: 0;
}
.e {
  list-style: none;
}


<ul data-testid="nav-container" class="a b c d e nav-container">
  <li data-testid="nav-item" class="nav-item">Hello</li>
</ul>
;
"
`;

exports[`components/Nav/NavContainer should render a NavContainer as a nav 1`] = `
".a {
  display: flex;
}
.b {
  flex-wrap: wrap;
}
.c {
  padding-left: 0;
}
.d {
  margin-bottom: 0;
}
.e {
  list-style: none;
}


<nav data-testid="nav-container" class="a b c d e nav-container">
  <li data-testid="nav-item" class="nav-item">Hello</li>
</nav>
;
"
`;

exports[`components/Nav/NavContainer should render a NavContainer of variant default 1`] = `
".a {
  display: flex;
}
.b {
  flex-wrap: wrap;
}
.c {
  padding-left: 0;
}
.d {
  margin-bottom: 0;
}
.e {
  list-style: none;
}


<ul data-testid="nav-container" class="a b c d e nav-container">
  <li data-testid="nav-item" class="nav-item">Hello</li>
</ul>
;
"
`;

exports[`components/Nav/NavContainer should render a NavContainer of variant pills 1`] = `
".a {
  display: flex;
}
.b {
  flex-wrap: wrap;
}
.c {
  padding-left: 0;
}
.d {
  margin-bottom: 0;
}
.e {
  list-style: none;
}


<ul data-testid="nav-container" class="a b c d e nav-container">
  <li data-testid="nav-item" class="nav-item">Hello</li>
</ul>
;
"
`;

exports[`components/Nav/NavContainer should render a NavContainer of variant tabs 1`] = `
".a {
  display: flex;
}
.b {
  flex-wrap: wrap;
}
.c {
  padding-left: 0;
}
.d {
  margin-bottom: 0;
}
.e {
  list-style: none;
}
.f {
  border-bottom-width: 1px;
}
.g {
  border-bottom-style: solid;
}
.h {
  border-bottom-color: #dee2e6;
}


<ul data-testid="nav-container" class="a b c d e f g h nav-container">
  <li data-testid="nav-item" class="nav-item">Hello</li>
</ul>
;
"
`;

exports[`components/Nav/NavContainer should render a NavContainer of variant underline 1`] = `
".a {
  display: flex;
}
.b {
  flex-wrap: wrap;
}
.c {
  padding-left: 0;
}
.d {
  margin-bottom: 0;
}
.e {
  list-style: none;
}
.f {
  row-gap: 1rem;
}
.g {
  column-gap: 1rem;
}


<ul data-testid="nav-container" class="a b c d e f g nav-container">
  <li data-testid="nav-item" class="nav-item">Hello</li>
</ul>
;
"
`;

exports[`components/Nav/NavContainer should render a NavContainer of variant vertical 1`] = `
".a {
  display: flex;
}
.b {
  flex-wrap: wrap;
}
.c {
  padding-left: 0;
}
.d {
  margin-bottom: 0;
}
.e {
  list-style: none;
}
.f {
  flex-direction: column;
}


<ul data-testid="nav-container" class="a b c d e f nav-container">
  <li data-testid="nav-item" class="nav-item">Hello</li>
</ul>
;
"
`;

exports[`components/Nav/NavContainer should render a centered NavContainer 1`] = `
".a {
  display: flex;
}
.b {
  flex-wrap: wrap;
}
.c {
  padding-left: 0;
}
.d {
  margin-bottom: 0;
}
.e {
  list-style: none;
}
.f {
  justify-content: center;
}


<ul data-testid="nav-container" class="a b c d e f nav-container">
  <li data-testid="nav-item" class="nav-item">Hello</li>
</ul>
;
"
`;

exports[`components/Nav/NavContainer should render a centered NavContainer of variant default 1`] = `
".a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}
.b {
  flex-wrap: wrap;
}
.c {
  padding-left: 0;
}
.d {
  margin-bottom: 0;
}
.e {
  list-style: none;
}
.f {
  justify-content: center;
}


<ul data-testid="nav-container" class="a b c d e f nav-container">
  <li data-testid="nav-item" class="nav-item">Hello</li>
</ul>
;
"
`;

exports[`components/Nav/NavContainer should render a centered NavContainer of variant pills 1`] = `
".a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}
.b {
  flex-wrap: wrap;
}
.c {
  padding-left: 0;
}
.d {
  margin-bottom: 0;
}
.e {
  list-style: none;
}
.f {
  justify-content: center;
}


<ul data-testid="nav-container" class="a b c d e f nav-container">
  <li data-testid="nav-item" class="nav-item">Hello</li>
</ul>
;
"
`;

exports[`components/Nav/NavContainer should render a centered NavContainer of variant tabs 1`] = `
".a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}
.b {
  flex-wrap: wrap;
}
.c {
  padding-left: 0;
}
.d {
  margin-bottom: 0;
}
.e {
  list-style: none;
}
.f {
  justify-content: center;
}
.g {
  border-bottom-width: 1px;
}
.h {
  border-bottom-style: solid;
}
.i {
  border-bottom-color: #dee2e6;
}


<ul data-testid="nav-container" class="a b c d e f g h i nav-container">
  <li data-testid="nav-item" class="nav-item">Hello</li>
</ul>
;
"
`;

exports[`components/Nav/NavContainer should render a centered NavContainer of variant underline 1`] = `
".a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}
.b {
  flex-wrap: wrap;
}
.c {
  padding-left: 0;
}
.d {
  margin-bottom: 0;
}
.e {
  list-style: none;
}
.f {
  justify-content: center;
}
.g {
  row-gap: 1rem;
}
.h {
  column-gap: 1rem;
}


<ul data-testid="nav-container" class="a b c d e f g h nav-container">
  <li data-testid="nav-item" class="nav-item">Hello</li>
</ul>
;
"
`;
